kernel/sched.c

```diff
@@ -735,16 +762,25 @@ static void z_thread_halt(struct k_thread *thread, k_spinlock_key_t key,
 
 			/* Now we know it's halting, but not necessarily
 			 * halted (suspended or aborted). Wait for the switch
-			 * to happen!
+			 * to happen unless deadlock was detected. If deadlock
+			 * was detected, we know we are in a special case where
+			 * all relevant CPUs are in ISRs trying to halt one or
+			 * more threads and no context switch can be done
+			 * until one of the ISRs exit.
 			 */
 			key = k_spin_lock(&sched_spinlock);
-			z_sched_switch_spin(thread);
```

Comment:
This looks wrong to me: if you're grabbing the scheduler lock here, then that means it was released somewhere above, after the entry to the function (which starts locked).  If that happened, then the CPU running the target thread **may not be `cpu` anymore**, right?  All use of that returned variable needs to happen before the lock is ever released.

Reply:
Pondering this one more ...

Reply:
Good catch. In this version, I have restricted to the creation of the _spinning on_ linked list to CPUs that are in an ISR. Also, if the other CPU is no longer executing `thread`, then it should mean that the target thread has been both halted and swapped out, which should address the issue you raised.